model/shared/src/main/scala/explore/data/tree/KeyedIndexedTree.scala

```diff
@@ -93,15 +95,55 @@ case class KeyedIndexedTree[K: Eq, A] private (
     KeyedIndexedTree(buildKeyMap(newTree, getKey), newTree)(getKey)
   }
 
+  def updated(key: K, newNode: A, newIndex: Index[K]): KeyedIndexedTree[K, A] = {
+    val current = getNodeAndIndexByKey(key)
+    current match
+      case None             => inserted(key, Node(newNode, Nil), newIndex)
+      case Some((value, _)) =>
+        removed(key).inserted(key, Node(newNode, value.children), newIndex)
+  }
+
   def collect[B](pf: PartialFunction[(K, Node[A], Index[K]), B]): List[B] =
     byKey
       .collect(pf.compose { case (k, node) => (k, node.map(_.elem), node.value.index) })
       .toList
 
+  def parentKeys(key: K): List[K] = {
+    @tailrec
+    def go(
+      acc:     List[K],
+      current: K
+    ): List[K] =
+      byKey.get(current) match
+        // Node not found, shouldn't happen 🤷
+        case None                                                     => acc
+        // We've found the 'root' node, so we're done
+        case Some(Node(IndexedElem(_, Index(None, _)), _))            => acc
+        // We've found a parent node, so we add it to the list and continue
+        case Some(Node(IndexedElem(_, Index(Some(parentKey), _)), _)) =>
+          go(parentKey +: acc, parentKey)
+    go(Nil, key)
+  }
+
+  def mapElement[B](f: A => B, getKey: B => K): KeyedIndexedTree[K, B] = {
+    val newTree: Tree[IndexedElem[K, B]] = tree.map { case IndexedElem(a, idx) =>
+      IndexedElem(f(a), idx)
+    }
+
+    def mapNode(node: Node[IndexedElem[K, A]]): Node[IndexedElem[K, B]] =
+      Node(IndexedElem(f(node.value.elem), node.value.index), node.children.map(mapNode))
+
+    val newByKey = byKey.map((k, node) => (k, mapNode(node)))
+    KeyedIndexedTree(newByKey, newTree)(getKey)
+  }
+
 }
 
 object KeyedIndexedTree {
   case class Index[K](parentKey: Option[K], childPos: Int)
+  object Index:
+    inline def apply[K](parentKey: Option[K], childPos: NonNegShort): Index[K] =
```

Comment:
Maybe it makes sense to change `childPos` in the class to a non-negative type. Not sure about it being a `Short` though. Is that how it comes from the ODB for groups?

Reply:
Yeah I thought about this too. But because there are multiple operations done on the childPos when updating the tree you'd have to do a lot of `NonNegShort.unsafeFrom(i.toShort)` which is kind of ugly. I agree it makes sense for it to be NonNegX though.

`NonNegShort` comes from the odb types for indices, yes